### PR TITLE
Add Fedora 36 to the RID graph

### PR DESCRIPTION
**Pull request by omajid**
Fedora 36 is under development:

    $ podman run -it registry.fedoraproject.org/fedora:rawhide /bin/cat /etc/os-release
    NAME="Fedora Linux"
    VERSION="36 (Container Image Prerelease)"
    ID=fedora
    VERSION_ID=36
    VERSION_CODENAME=""
    PLATFORM_ID="platform:f36"
    PRETTY_NAME="Fedora Linux 36 (Container Image Prerelease)"
    ANSI_COLOR="0;38;2;60;110;180"
    LOGO=fedora-logo-icon
    CPE_NAME="cpe:/o:fedoraproject:fedora:36"
    HOME_URL="https://fedoraproject.org/"
    DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/rawhide/system-administrators-guide/"
    SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
    BUG_REPORT_URL="https://bugzilla.redhat.com/"
    REDHAT_BUGZILLA_PRODUCT="Fedora"
    REDHAT_BUGZILLA_PRODUCT_VERSION=rawhide
    REDHAT_SUPPORT_PRODUCT="Fedora"
    REDHAT_SUPPORT_PRODUCT_VERSION=rawhide
    PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
    VARIANT="Container Image"
    VARIANT_ID=container

I couldn't get the changes runtime.json to show up until I built with a customtarget. I updated readme.md with that extra information too.

### CODE REVIEW

**Comment by omajid** [comment]
Hey @ViktorHofer am I doing something wrong with `/t:UpdateRuntimeJson`? Should it have worked out of the box?

File: src/libraries/Microsoft.NETCore.Platforms/readme.md [hunk 1]
```diff
@@ -100,7 +100,7 @@ For example:
 
 This will create a new RID for `myLinuxDistro` where `myLinuxDistro` should be the string used for the `ID=` value in the `/etc/os-release` file.
 
-Whenever modifying the `runtimeGroups.props` you should rebuild the project with `/p:UpdateRuntimeFiles=true` so that your changes will be regenerated in the checked-in `runtime.json`.
+Whenever modifying the `runtimeGroups.props` you should rebuild the project with `/t:UpdateRuntimeJson /p:UpdateRuntimeFiles=true` so that your changes will be regenerated in the checked-in `runtime.json`.
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.NETCore.Platforms/readme.md:103]: Now that we have a fairly an easy way to produce the package (building the Platforms packages doesn't even require any up-front build), I would recommend to create the package and diff its inputs. The `UpdateRuntimeJson` target is already hooked to run before packaging, so invoking that manually shouldn't be necessary anymore.

```suggestion
Whenever modifying the `runtimeGroups.props` make sure to pack the project via the `dotnet pack` command and inspect if the generated package contains the desired changes.
```

**Comment by ViktorHofer** [comment]
> Hey @ViktorHofer am I doing something wrong with /t:UpdateRuntimeJson? Should it have worked out of the box?

I don't know what UpdateRuntimeJson is exactly doing without looking into its sources. I think @ericstj should be able to answer.

**Comment by ericstj** [comment]
> am I doing something wrong with /t:UpdateRuntimeJson? Should it have worked out of the box?

It would have when running pack.  If you aren't packing you can run the target directly.  I think the docs are unclear about this since they say "rebuild the project" which made sense when it was a dedicated pkgproj but is no longer the best wording now that it's a CSProj.  I'm fine with the update you made to the docs.

**Review by ViktorHofer — APPROVED** [review]

**Comment by ViktorHofer** [comment]
Merging as the previous run was successful and the recent commit just update a docs file.

Thanks a lot @omajid

**Comment by tmds** [comment]
@ericstj @ViktorHofer I thought, thanks to https://github.com/dotnet/runtime/pull/50818, we no longer had to make these changes?

**Comment by ViktorHofer** [comment]
You likely want what's proposed in https://github.com/NuGet/Home/issues/5862 as well. https://github.com/dotnet/runtime/pull/50818 helps with source build as you can define a new RID via a variable but I don't see how this would help when not source building.

**Comment by tmds** [comment]
> #50818 helps with source build

Yes.

> I don't see how this would help when not source building.

I'm not aware of issues.
If there are, probably they can be addressed to make this type change also unnecessary for non-source build.

**Comment by ViktorHofer** [comment]
> In practice, we've seen this is not scalable and it can cause applications to fail when a system is upgraded from one version to the next.

My assumption is that the current design isn't scalable because of the issues listed in https://github.com/NuGet/Home/issues/5862.

cc @ericstj for his opinion

**Comment by ericstj** [comment]
https://github.com/dotnet/runtime/pull/50818 helps folks avoid source changes when building dotnet/runtime on new distros.  It doesn't eliminate the need to have discrete RIDs listed in the RID graph in order to support those new distros at runtime/design-time.  The Host, NuGet, and SDK still depend on that.   I'd love to see us eliminate that in v.Next.  cc @richlander

**Comment by richlander** [comment]
My most frequent experience with this is with Alpine, which ships frequently. We're almost always waiting on new RIDs to release a new Alpine. That makes little sense to me.

I think the order should be (using existing versions as examples):

- Release new .NET 5 container images for Alpine 3.14, ASAP.
- Declare somehow that Alpine 3.14 is equivalent to Alpine 3.13 (which has an existing RID)
- This means we don't have to wait on ANY product changes.
- At our convenience (within 2 months), add Alpine 3.14 to the RID graph so that people can explicitly target it.

Perhaps we could do this today, but it's not what we do.

/cc @mthalman

**Comment by tmds** [comment]
> At our convenience (within 2 months), add Alpine 3.14 to the RID graph so that people can explicitly target it.

For distros that move fast like Fedora and Alpine, we continuously need to upgrade the rid graph.
As I mentioned earlier, it is not only a scalability issue for us updating the graph, but also the same scalability issue exists for users targeting these individual rids.

That makes it impractical to use these rids, and a preferable solution is to do assets selection at runtime using NativeLibrary, RuntimeInformation.RuntimeIdentifier, ...

It may still be interesting to target the parent rid (e.g. `fedora-x64`).
We can extend the Host, NuGet, and SDK, ... so they have some logic like https://github.com/dotnet/runtime/pull/50818.
That means when .NET runs on the unknown `alpine.3.14-x64`, it will still consider `alpine-x64` assets.

@richlander @ericstj your thoughts?

**Pull request merged**